File: testsuite/src/java/org/xmlBlaster/test/client/TestRequestResponse.java
/**
 * 
 */
package org.xmlBlaster.test.client;

import java.util.logging.Logger;

import junit.framework.TestCase;

import org.xmlBlaster.client.I_Callback;
import org.xmlBlaster.client.I_XmlBlasterAccess;
import org.xmlBlaster.client.key.PublishKey;
import org.xmlBlaster.client.key.UpdateKey;
import org.xmlBlaster.client.qos.ConnectQos;
import org.xmlBlaster.client.qos.ConnectReturnQos;
import org.xmlBlaster.client.qos.DisconnectQos;
import org.xmlBlaster.client.qos.PublishQos;
import org.xmlBlaster.client.qos.PublishReturnQos;
import org.xmlBlaster.client.qos.UpdateQos;
import org.xmlBlaster.test.Util;
import org.xmlBlaster.util.EmbeddedXmlBlaster;
import org.xmlBlaster.util.Global;
import org.xmlBlaster.util.MsgUnit;
import org.xmlBlaster.util.SessionName;
import org.xmlBlaster.util.XmlBlasterException;
import org.xmlBlaster.util.def.Constants;
import org.xmlBlaster.util.qos.address.Destination;

/**
 * Testing XmlBlasterAccess.request() method.
 * <br /> 
 * We send a request and expect a response.
 * @author Marcel Ruff
 */
public class TestRequestResponse extends TestCase {
   private static Logger log = Logger.getLogger(TestRequestResponse.class.getName());

   private Global glob;
   private EmbeddedXmlBlaster serverThread;
   private int serverPort = 9583;
   private boolean startEmbedded = true;
   
   private I_XmlBlasterAccess sender = null;
   private final String senderName = "TheDesperate";
   private I_XmlBlasterAccess receiver = null;
   private final String receiverName = "TheKnowing";

   public void testRequestResponse() {
      Global glob = Global.instance();
      
      try {

         {  // setup the sender client ...
            sender = glob.getXmlBlasterAccess();

            ConnectQos qos = new ConnectQos(sender.getGlobal(), senderName, "secret");
            ConnectReturnQos conRetQos = sender.connect(qos, null); // Login to xmlBlaster

            log.info("Sender connected to xmlBlaster " + conRetQos.getSessionName().getRelativeName());
         }


         {  // setup the receiver client which processes the request (usually another process) ...
            Global globReceiver = glob.getClone(null);
            receiver = globReceiver.getXmlBlasterAccess();

            ConnectQos qos = new ConnectQos(receiver.getGlobal(), receiverName, "secret");
            ConnectReturnQos conRetQos = receiver.connect(qos, new I_Callback() {
               public String update(String cbSessionId, UpdateKey updateKey, byte[] content, UpdateQos updateQos) {
                  log.info(receiverName+": Receiving asynchronous message '" + updateKey.getOid() + "' in receiver default handler");
                  log.info(receiverName+": Received: " + updateKey.toXml() + "\n <content>" + new String(content) + "</content>" + updateQos.toXml());

                  if (updateKey.isInternal()) return "";
                  if (updateQos.isErased()) return "";
                  try {
                     String tempTopicOid = updateQos.getClientProperty(Constants.JMS_REPLY_TO, ""); // __jms:JMSReplyTo
                     log.info(receiverName+": Got request, using topic '" + tempTopicOid + "' for response");

                     // Send reply back ...
                     PublishKey pk = new PublishKey(receiver.getGlobal(), tempTopicOid, "text/plain", "1.0");
                     PublishQos pq = new PublishQos(receiver.getGlobal());
                     MsgUnit msgUnit = new MsgUnit(pk, "On doubt no ultimate truth, my dear.", pq);
                     //try { Thread.sleep(8000); } catch (InterruptedException e) { e.printStackTrace(); }
                     PublishReturnQos retQos = receiver.publish(msgUnit);
                     log.info(receiverName+": Published reply message using temporary topic " + retQos.getKeyOid());
                  }
                  catch (XmlBlasterException e) {
                     log.severe(receiverName+": Sending reply to " + updateQos.getSender() + " failed: " + e.getMessage());
                  }
                  return "";
               }
            });  // Login to xmlBlaster, default handler for updates

            log.info("Receiver connected to xmlBlaster " + conRetQos.getSessionName().getRelativeName());
         }

         // Send a message to 'receiver' and block for the reply
         PublishKey pk = new PublishKey(sender.getGlobal(), "requestForEnlightenment");
         PublishQos pq = new PublishQos(sender.getGlobal());
         pq.addDestination(new Destination(new SessionName(sender.getGlobal(), receiverName)));
         MsgUnit msgUnit = new MsgUnit(pk, "Tell me the truth!", pq);
         MsgUnit[] replies = sender.request(msgUnit, 3000, 1);
         assertEquals(senderName+": Missing reply message.", 1, replies.length);
         log.info(senderName+": Got " + replies.length + " reply :\n" + replies[0].toXml());
      }
      catch (XmlBlasterException e) {
         fail("We have a problem: " + e.getMessage());
      }
      finally {
         if (sender != null && sender.isConnected()) { sender.disconnect(new DisconnectQos(sender.getGlobal())); }
         if (receiver != null && receiver.isConnected()) { receiver.disconnect(new DisconnectQos(receiver.getGlobal())); }
      }
   }

   /**
    * @param arg0
    */
   public TestRequestResponse(String arg0) {
      super(arg0);
   }

   /* (non-Javadoc)
    * @see junit.framework.TestCase#setUp()
    */
   protected void setUp() throws Exception {
      super.setUp();
      this.glob = Global.instance();
      this.startEmbedded = this.glob.getProperty().get("startEmbedded", this.startEmbedded);
      if (this.startEmbedded) {
         glob.init(Util.getOtherServerPorts(serverPort));
         String[] args = { };
         glob.init(args);
         this.serverThread = EmbeddedXmlBlaster.startXmlBlaster(glob);
         log.info("XmlBlaster is ready for testing");
      }

   }

   /* (non-Javadoc)
    * @see junit.framework.TestCase#tearDown()
    */
   protected void tearDown() throws Exception {
      super.tearDown();
      if (this.startEmbedded) {
         try { Thread.sleep(500L); } catch( InterruptedException i) {} // Wait some time
         EmbeddedXmlBlaster.stopXmlBlaster(this.serverThread);
         this.serverThread = null;
      }

      // reset to default server port (necessary if other tests follow in the same JVM).
      Util.resetPorts();
   }
}

File: testsuite/src/c++/TestRequestResponse.cpp
/*-----------------------------------------------------------------------------
Name:      TestRequestResponse.cpp
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
Comment:   Demo code for a client using xmlBlaster
           build -DexeName=TestRequestResponse cpp-test-single
-----------------------------------------------------------------------------*/
#include "TestSuite.h"
#include <iostream>

/**
 * This client tests the request/reply pattern
 */

using namespace std;
using namespace org::xmlBlaster::util;
using namespace org::xmlBlaster::util::qos;
using namespace org::xmlBlaster::util::thread;
using namespace org::xmlBlaster::client;
using namespace org::xmlBlaster::client::qos;
using namespace org::xmlBlaster::client::key;
using namespace org::xmlBlaster::authentication;

namespace org { namespace xmlBlaster { namespace test {

/**
 * Callback implementation of the receiver client. 
 */
class ReceiverCallback : public I_Callback {
private:
   XmlBlasterAccess &receiver_;
   string name_;
   I_Log& log_;

public:
   ReceiverCallback(XmlBlasterAccess &receiver, I_Log& log, const string& name) : receiver_(receiver), log_(log) {
      name_ = name;
   }

   string update(const string &sessionId,
               UpdateKey &updateKey,
               const unsigned char * /*content*/, long /*contentSize*/,
               UpdateQos &updateQos) 
   {
      log_.info("update", string("Receiving update on callback '") + name_ + "' of message oid=" +
                updateKey.getOid() + " state=" + updateQos.getState() +
                " authentication sessionId=" + sessionId + " ...");

      //if (updateKey.isInternal()) return "";
      if (updateQos.isErased()) return "";
      
      try {
         string tempTopicOid = updateQos.getClientProperty(Constants::JMS_REPLY_TO, string("")); // __jms:JMSReplyTo
         log_.info("update", name_+": Got request, using topic '" + tempTopicOid + "' for response");

         // Send reply back ...
         PublishKey pk(receiver_.getGlobal(), tempTopicOid, "text/plain", "1.0");
         PublishQos pq(receiver_.getGlobal());
         MessageUnit msgUnit(pk, "On doubt no ultimate truth, my dear.", pq);
         PublishReturnQos retQos = receiver_.publish(msgUnit);
         log_.info("update", name_+": Published reply message using temporary topic " + retQos.getKeyOid());
      }
      catch (XmlBlasterException &e) {
         log_.error("update", string(name_)+": Sending reply to " + updateQos.getSender()->getAbsoluteName() + " failed: " + e.getMessage());
      }
      return "<qos><state id='OK'/></qos>";
   }
};


class TestRequestResponse: public TestSuite, public virtual I_Callback 
{
private:
   bool   messageArrived_;      // = false;
   string subscribeOid_;
   string publishOid_;          // = "dummy";
   string senderName_;
   string receiverName_;        // sender/receiver is here the same client
   ConnectReturnQos returnQos_;
   ReceiverCallback *cbReceiver;

   GlobalRef globalReceiver_;
   XmlBlasterAccess receiver_;

   /**
    * Constructs the TestRequestResponse object.
    * <p />
    * @param testName  The name used in the test suite
    * @param loginName The name to login to the xmlBlaster
    */
 public:
   TestRequestResponse(int args, char *argc[], const string &senderName, const string &receiverName)
      : TestSuite(args, argc, "TestRequestResponse"),
        returnQos_(global_),
        globalReceiver_(global_.createInstance("receiverGlobal")),
	receiver_(globalReceiver_)
   {
      globalReceiver_->initialize(args, argc);
      senderName_          = senderName;
      receiverName_        = receiverName;
      publishOid_          = "dummy";
      cbReceiver = new ReceiverCallback(receiver_, log_, "callbackReceiver");
   }

   virtual ~TestRequestResponse() 
   {
      delete cbReceiver;
   }

   /**
    * Sets up the fixture. <p />
    * Connect to xmlBlaster and login
    */
   void setUp() 
   {
      log_.info(ME, "Trying to connect to xmlBlaster with C++ client lib " + Global::getVersion() + " from " + Global::getBuildTimestamp());
      TestSuite::setUp();
      const string passwd = "secret";
      try {
         ConnectQos connQos(global_, senderName_);
         returnQos_ = connection_.connect(connQos, this);
         log_.info(ME, string("connected '") + senderName_ + "'");
      }
      catch (XmlBlasterException &e) {
         log_.error(ME, string("Login failed: ") + e.toXml());
         usage();
         assert(0);
      }

      try {
         ConnectQos connQos(receiver_.getGlobal(), receiverName_);
         returnQos_ = receiver_.connect(connQos, cbReceiver);
         log_.info(ME, string("connected '") + receiverName_ + "'");
      }
      catch (XmlBlasterException &e) {
         log_.error(ME, string("Login failed: ") + e.toXml());
         usage();
         assert(0);
      }
   }


   /**
    * Tears down the fixture. <p />
    * cleaning up .... erase() the previous message OID and logout
    */
   void tearDown() 
   {
      log_.info(ME, "Cleaning up test");
      connection_.disconnect(DisconnectQos(global_));
      receiver_.disconnect(DisconnectQos(receiver_.getGlobal()));
      TestSuite::tearDown();
   }

   /**
    * TEST
    */
   void testRequest() 
   {
      if (log_.trace()) log_.trace(ME, "Publishing a message (old style) ...");
      PublishKey pubKey(global_, "requestForEnlightenment");
      PublishQos pubQos(global_);
      Destination destination(global_, receiverName_);
      pubQos.addDestination(destination);
      MessageUnit msgUnit(pubKey, "Tell me the truth!", pubQos);
      try {
         std::vector<org::xmlBlaster::util::MessageUnit> replies = 
	    connection_.request(msgUnit, 6000, 1);
	 assertEquals(log_, ME, 1, replies.size(), "missing response");
         log_.info(ME, senderName_+": Got " + lexical_cast<string>(replies.size()) + " reply :\n" + replies[0].toXml());
         log_.info(ME, replies[0].getContentStr() + " size=" + lexical_cast<string>(replies[0].getContentLen()));
	 assertEquals(log_, ME, "On doubt no ultimate truth, my dear.", replies[0].getContentStr(), "response from a wise man");
      }
      catch(XmlBlasterException &e) {
         log_.warn(ME, string("XmlBlasterException: ")+e.toXml());
         assert(0);
      }
   }

   string update(const string &sessionId,
               UpdateKey &updateKey,
               const unsigned char *content, long contentSize,
               UpdateQos &updateQos) 
   {
      log_.error(ME, string("Unexpected receiving update of message oid=") +
                updateKey.getOid() + " state=" + updateQos.getState() +
                " authentication sessionId=" + sessionId + " ...");
      return "<qos><state id='OK'/></qos>";
   }


   /**
    * Little helper, waits until the variable 'messageArrive' is set
    * to true, or returns when the given timeout occurs.
    * @param timeout in milliseconds
    */
private:
   void usage() const
   {
      TestSuite::usage();
      log_.plain(ME, "----------------------------------------------------------");
      log_.plain(ME, "Testing C++ request/reply pattern");
      log_.plain(ME, "Usage:");
      XmlBlasterAccess::usage();
      log_.usage();
      log_.plain(ME, "Example:");
      log_.plain(ME, "   TestRequestResponse -trace true");
      log_.plain(ME, "----------------------------------------------------------");
   }
};

}}} // namespace

using namespace org::xmlBlaster::test;

int main(int args, char *argc[]) 
{
   try {
      org::xmlBlaster::util::Object_Lifetime_Manager::init();
      TestRequestResponse testReq(args, argc, "TheDesperate", "TheKnowing");
 
      testReq.setUp();
      testReq.testRequest();
      testReq.tearDown();

      Thread::sleepSecs(1);
   }
   catch (XmlBlasterException& ex) {
      std::cout << ex.toXml() << std::endl;
   }
   catch (bad_exception& ex) {
      cout << "bad_exception: " << ex.what() << endl;
   }
   catch (exception& ex) {
      cout << " exception: " << ex.what() << endl;
   }
   catch (string& ex) {
      cout << "string: " << ex << endl;
   }
   catch (char* ex) {
      cout << "char* :  " << ex << endl;
   }

   catch (...)
   {
      cout << "unknown exception occured" << endl;
      XmlBlasterException e(INTERNAL_UNKNOWN, "main", "main thread");
      cout << e.toXml() << endl;
   }

   org::xmlBlaster::util::Object_Lifetime_Manager::fini();
   return 0;
}


